### ISSUE TITLE
ckSurf overrides ALL Admin-Tag plugins.

### ISSUE THREAD

**Issue by Sandman007** [nikooo777/ckSurf#57]
Hello,

First off I know this isn't your fault. This has been an issue since the original ckSurf plugin. But can you either completely remove the Admin tags function or change they way it handles tags?

Even when I set AdminTags to 0 in the cfg file, this plugin break compatibility with every AdminTags plugin out there that modifies tags in the scoreboard. I have tried even some very recent plugins on alliedmodders like these:

https://forums.alliedmods.net/showpost.php?p=2421587&postcount=61
https://forums.alliedmods.net/showthread.php?t=293563
https://forums.alliedmods.net/showthread.php?t=241345&page=5

All of these get overridden by ckSurf. You might see your custom tag from the above plugins for a like a split second when loading into the map but then ckSurf takes over.

Thanks!

**Comment by M-archand** [comment]
I agree, I would pay good money for a admin/other tag (scoreboard+chat) plugin that is compatible with ckSurf, or a rewrite of ckSurf itself that integrates a more comprehensive "admin" tag setup..

**Comment by 2called-chaos** [comment]
@M-archand can you define more comprehensive?

The reason why `AdminTags` doesn't do anything for your problem is because this setting only specifies wether admins have the `ADMIN` tag, there is still skill groups, country indicator and custom titles taking place.

Since there is so much scoreboard magic going on I don't think there is an easy way to make it compatible with other plugins (unless they all follow some sort of standardized API to get the rank titles). Unless you want no scoreboard manipulation at all, which makes other plugins work but you will loose _ALL_ ckSurf features (as mentioned above).

I personally use the custom rank feature to give myself the owner tag (my friend also has "z" flag but he doesn't get to have this tag :P).

Alternatively, one could try to alter it here (to distinguish between different flags): https://github.com/nikooo777/ckSurf/blob/master/csgo/addons/sourcemod/scripting/ckSurf/misc.sp#L2129

Note: There is a lot of code that checks whether the clantag contains admin and I honestly don't understand why.

I also would be interested in rewriting ckSurf because the codebase is horribly, horribly bad but I'm not good enough with sourcemod I'm afraid. Currently I'm trying to fix things but the codebase doesn't make it easy.

**Comment by M-archand** [comment]
@2called-chaos by more comprehensive I guess I mean that so it's not so one dimensional by only checking for the 1 defined flag in the cfg and then giving the one available title "ADMIN". 

In my perfect world you'd be able to set a **hierarchy** of roles based on flag checks (mainly custom 1-6), then assign different tags: owner, admin, mod, vip, donator, thing1, name1, etc, etc... all within ckSurf

Like you said, there is so much going on with chat/scoreboard manipulation in this plugin, so when you use 3rd party tag plugins you end up with crap like this happening, where stuff double writes or just doesn't work at all:
![68747470733a2f2f692e696d6775722e636f6d2f6c6d43577534482e706e67](https://user-images.githubusercontent.com/13469995/31318906-0370017a-ac30-11e7-8773-7ec715a43396.png)

Like you had also mentioned: [L2131-L2135](https://github.com/nikooo777/ckSurf/blob/c1fa42c671ef7c9e26d19f8c1febb688da4d4511/csgo/addons/sourcemod/scripting/ckSurf/misc.sp#L2131-L2135) could be altered, just replicating it and throwing in some `else` statements. I'm sure I could do some trial and error and see what comes from it, but I don't know much about SP or C, so if there's a bunch of other checks or dependents/precedents then I'd probably get lost within the 1000's of lines of code spanning across different files...

Right now I'm in the same boat as you, using custom skill groups to give special tags like owner, but there's a lot of features I'd like to eventually implement to my surf servers where it just wouldn't be feasible to assign custom skill groups manually. It could also be approached by having the custom skill groups be flag based like how "VIP" is with `ck_autovip_flag`

**Comment by M-archand** [comment]
Tested this out quickly and got it to work a little bit. Adding another manipulated `else if` statement after the lines referenced above I got multiple tags to work at once.

```
	// Admin Clantag
	if (GetConVarBool(g_hAdminClantag))
	{ if (GetUserFlagBits(client) & ADMFLAG_ROOT || GetUserFlagBits(client) & ADMFLAG_GENERIC)
		{
			Format(g_pr_chat_coloredrank[client], 128, "%s %cADMIN%c", g_pr_chat_coloredrank[client], LIMEGREEN, WHITE);
			Format(g_pr_rankname[client], 128, "ADMIN");
			return;
		}
		else if (GetUserFlagBits(client) & ADMFLAG_ROOT || GetUserFlagBits(client) & ADMFLAG_CUSTOM1)
		{
			Format(g_pr_chat_coloredrank[client], 128, "%s %cDONOR%c", g_pr_chat_coloredrank[client], LIMEGREEN, WHITE);
			Format(g_pr_rankname[client], 128, "DONOR");
			return;
		}
	}
```

So if someone has "o" flag for example they'll now get a "DONOR" tag. Seems to work on both scoreboard & chat. Examples below: 
![j4chab7](https://user-images.githubusercontent.com/13469995/31321371-93b5a778-ac5b-11e7-8450-9f922bd079d5.png)
![fzvkepd](https://user-images.githubusercontent.com/13469995/31321384-da7ef92a-ac5b-11e7-8837-9f8c9f05d338.png)

Going to mess around with this a little more, but I really know shit all about SP so if anyone knows how to incorporate this into the cksurf.cfg to easily change flags/tags besides recompile that would be amazing

**Comment by Sandman007** [comment]
Good Job!  I am gonna add this and recompile on my end as well. I am learning SP but I do not know enough yet myself to implement this into a CFG file. I may try tomorrow tho. Nice work!

**Comment by 2called-chaos** [comment]
I added one feature once that had a setting so maybe this may help you on your journey :)

https://github.com/nikooo777/ckSurf/pull/31/files

Personally I don't really buy into the idea with the flags though. How would you want to configure it? Use the order in which you put it into the config (e.g. root should be checked first, then ban flag, ...). How to configure "or's" and/or "and's". I guess it would be more powerful to enhance the current custom rank stuff (make it DB based and remove the rank limit) + enhancing the autovip feature as to autogrant titles on generic flags.

One reason why I use custom ranks even for generic admins is that with custom titles, everyone has the option to wear that rank or disable/choose a different one with `/titles`, with the convar it's basically enforced for everyone (which might be desired, but I wouldn't want it for example). Don't know what the general consensus is there.

**Comment by M-archand** [comment]
I agree with your points about issues with flags. Some people might not want certain titles to be shown, I know it gets kind of annoying for myself/admins who are put under a bit of a microscope when they're on servers due to forced tags. Then again I have had a lot of issues with people impersonating admins which led me to force tags so it won't happen anymore. However, maybe someone is a donor who doesn't want that tag (they don't want the attention). 

I just think flags are a good may to do anything because you can have so many other plugins doing really cool things that are automated. For example, SourceBans++ handles all donation flags on my website on a 30 day rotating basis giving certain flags, I also have another plugin that gives people  certain flags if they're subscribed to a certain Twitch channel (I have a partnership with them), and the list goes on. So if you can pair tags with flags I feel like it opens the door to a lot of things.

As for it being set in the custom skill groups, I know a lot of people might want a special tag like admin/donor or whatever, but still be able to show their surf rank that they've worked towards for maybe months. Also, only certain flags can give out titles in ckSurf right? This goes along with the manual input issue that I mentioned earlier in the thread, just don't have the time to assign this for people that aren't admins. 

I think the best option would be to have it set up very similar to [this plugin](https://forums.alliedmods.net/showthread.php?t=299733) in the sense that it gives you the ability to do !title, then have a menu show up (like ckSurf custom) with titles available to you based on what flags you have, where you get to choose which one you want, or if it's just turned off. Obviously that would be a complete rewrite of how the current admin tags work so I can't see it happening within the current spaghetti code

**Comment by fluffyst** [comment]
How about a key value system with a file similar to like

**Comment by fluffyst** [comment]
How about a key value system with a file similar to like

```
"flag" "title"
"z" "{darkred}Owner"
"a" {yellow}VIP"
```
